packages/cip30/src/CardanoWallet.ts

```diff
@@ -0,0 +1,167 @@
+
+/**
+ * CIP30 Specification version
+ */
+export type SpecificationVersion = string;
+
+/**
+ * Unique identifier, used to inject into the cardano namespace
+ */
+export type WalletName = string;
+
+/**
+ * This is the entrypoint to start communication with the user's wallet.
+ *
+ * The wallet should request the user's permission to connect the web page to the user's wallet,
+ * and if permission has been granted, the full API will be returned to the dApp to use.
+ *
+ * The wallet can choose to maintain a whitelist to not necessarily ask the user's permission
+ * every time access is requested, but this behavior is up to the wallet and should be transparent
+ * to web pages using this API.
+ *
+ * If a wallet is already connected this function should not request access a second time,
+ * and instead just return the API object.
+ *
+ * Errors: `ApiError`
+ */
+export type Enable = () => Promise<CardanoWalletApi>;
+
+/**
+ * Returns true if the dApp is already connected to the user's wallet, or if requesting access
+ * would return true without user confirmation (e.g. the dApp is whitelisted), and false otherwise.
+ *
+ * If this function returns true, then any subsequent calls to wallet.enable()
+ * during the current session should succeed and return the API object.
+ *
+ * Errors: `ApiError`
+ */
+export type IsEnabled = () => Promise<Boolean>;
+
+export type WalletProperties = { name: WalletName; version: SpecificationVersion };
+
+/**
+ * Resolve true to authorise access to the CardanoWalletAPI, or resolve false to deny.
+ *
+ * Errors: `ApiError`
+ */
+export type RequestAccess = () => Promise<boolean>;
+
+export type WalletOptions = {
+  logger?: Logger;
+  persistAllowList?: boolean;
+  storage?: Storage;
+};
+
+export class CardanoWallet {
+  readonly version: SpecificationVersion;
+  readonly name: WalletName;
+
+  private allowList: string[];
+  private logger: Logger;
+
+  constructor(
+    properties: WalletProperties,
+    private api: CardanoWalletApi,
+    private window: Window & { cardano?: Record<string, CardanoWallet> },
+    private requestAccess: RequestAccess,
+    private options?: WalletOptions
+  ) {
+    this.logger = options.logger ?? dummyLogger;
+    this.name = properties.name;
+    this.version = properties.version;
+
+    if (typeof options.persistAllowList === 'undefined') {
+      options.persistAllowList = false;
+    }
+
+    this.allowList = this.options.persistAllowList ? this.getAllowList() : [];
+
+    if (!this.window.cardano) {
+      this.logger.debug(
+        {
+          module: 'CardanoWallet',
+          walletName: this.name
+        },
+        'Creating cardano global scope'
+      );
+      this.window.cardano = {};
+    } else {
+      this.logger.debug(
+        {
+          module: 'CardanoWallet',
+          walletName: this.name
+        },
+        'Cardano global scope exists'
+      );
+    }
+
+    this.window.cardano[properties.name] = this.window.cardano[properties.name] || this;
```

Comment:
Returning the whole instance exposes the API and other class properties, so we need to return just an object with references to the instance `name`, `version`, `isEnabled`, and `enable`. It would be best to add a new interface `CardanoWalletPublicApi` in a separate file

Reply:
`CardanoWalletPublicApi` might be confusing with `CardanoWalletApi` ?

Reply:
But yeah I agree we should sort this.